RMextract/getIONEX.py

```diff
@@ -626,105 +613,120 @@ def get_urllib_IONEXfile(time="2012/03/23/02:20:10.01",
+                    logger.error(f"Primary and Backup Server not responding: {e}") #enable in lover environment
+    
+    
+    if isinstance(formatter, str):
+        try:
+            formatter = KNOWN_FORMATTERS[formatter]
+        except KeyError:
+            raise ValueError(f"Unknown formatter {formatter} - please provide a callable")
+    if formatter is None:
+        # Check known servers
+        for known_server in KNOWN_FORMATTERS.keys():
+            if known_server in server:
+                formatter = KNOWN_FORMATTERS.get(known_server)
+                break
+        if formatter is None:
+            raise ValueError(f"Unknown server {server} - please provide a formatter")
+    
+    url = formatter(server=server, prefix=prefix, year=year, dayofyear=dayofyear)
+
+    logger.info(f"Constructed {url=}.")
 
-    try:
-        primary = request.urlopen(server,timeout=30)
-        serverfound = True
-    except:
-            try:
-                secondary = request.urlopen(backupserver,timeout=30)
-                backupfound = True
-                server=backupserver
-                logging.warning(f"Primary IONEX host '{server}' resolution "
-                                            f"failure. Trying backup at '{backupserver}'")
-            except:
-                logging.error('Primary and Backup Server not responding') #enable in lover environment
-    if "http://ftp.aiub.unibe.ch" in server:
-        url = "http://ftp.aiub.unibe.ch/CODE/%4d/%s%03d0.%02dI.Z"%(year,prefix.upper(),dayofyear,yy)
-    elif "http://cddis.gsfc.nasa.gov" in server:
-        url = "http://cddis.gsfc.nasa.gov/gnss/products/ionex/%4d/%03d/%s%03d0.%02di.Z"%(year,dayofyear,prefix,dayofyear,yy)
-    elif "igsiono.uwm.edu.pl" in server:
-        url = "https://igsiono.uwm.edu.pl/data/ilt/%4d/igrg%03d0.%02di"%(year,dayofyear,yy)
 
     # Download IONEX file, make sure it is always uppercase
-    fname = outpath+'/'+(url.split('/')[-1]).upper()
-    try:
-        site = request.urlopen(url,timeout=30)
-    except:
-        logging.info("No files found on %s for %s",server,fname)
-        return -1
-
-    output=open(fname,'wb')
-    output.write(site.read())
-    output.close()
-    ###### gunzip files
-    if fname[-2:].upper()==".Z":
-        command = "gunzip -dc %s > %s" % (fname, fname[:-2])
-        retcode = os.system(command)
-        if retcode:
-            raise RuntimeError("Could not run '%s'" % command)
-        else:
-            os.remove(fname)
-        fname=fname[:-2]
-    #returns filename of uncompressed file
-    return fname
+    fname = outpath / Path(url).name.upper()
+    out_fname = fname.with_suffix("") if fname.suffix == ".Z" else fname    
+    
+    # First, if the final file already exists, simply return
+    if not overwrite and out_fname.exists():
+        return out_fname.as_posix()
+
+    # Here we actually download the file. Lets make sure though that this
+    # file is not downloaded by another task first
```

Comment:
Indeed this was because we hit race conditions during parallel execution. Although, I think this use-case is possibly unusual. Our current workaround is to execute a singular task, which initialises the files, then trigger the full parallel execution which is fine with the existence check.